src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+            self.device.set_adc_trigger_mode(1)
+            self.device.set_qunit_mode(1)
+            raw = self.device.start_qunit_acquisition(options.nshots, readout_qubits)
+            res = {qubit: SampleResults(states) for qubit, states in raw.items()}
+            for ro_pulse in readout_pulses:
+                res[ro_pulse.serial] = res[ro_pulse.qubit]
+            return res
+
+    def play_sequences(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequences: List[PulseSequence],
+        options: ExecutionParameters,
+    ):
+        """Wrapper for play."""
+        return [
+            self.play(qubits, couplers, sequence, options) for sequence in sequences
+        ]
+
+    def process_readout_signal(
+        self,
+        adc_raw_data: Dict[int, np.ndarray],
+        sequence: List[Pulse],
+        qubits: Dict[QubitId, Qubit],
+        options: ExecutionParameters,
+    ):
+        """Processes the raw signal from the ADC into IQ values."""
+
+        adc_sampling_rate = self.device.adc_sampling_rate * 1e6
+        t = np.arange(self.device.adc_sample_size) / adc_sampling_rate
+        results = {}
+
+        for readout_pulse in sequence:
+            qubit = qubits[readout_pulse.qubit]
+            _, adc = qubit.readout.ports
+
+            raw_signal = adc_raw_data[adc]
+            sin = np.sin(2 * np.pi * readout_pulse.frequency * t)
+            cos = np.sin(2 * np.pi * readout_pulse.frequency * t)
+
+            I = np.dot(raw_signal, cos)
+            Q = np.dot(raw_signal, sin)
+            results[readout_pulse.qubit] = IntegratedResults(I + 1j * Q)
+
+            if options.averaging_mode is not AveragingMode.SINGLESHOT:
+                results[readout_pulse.qubit] = results[readout_pulse.serial] = results[
+                    readout_pulse.qubit
+                ].average
+            else:
+                results[readout_pulse.serial] = results[readout_pulse.qubit]
+
+        return results
+
+    def sweep(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequence: PulseSequence,
+        options: ExecutionParameters,
+        *sweeper: Sweeper,
+    ):
+        # Record pulse values before sweeper modification
+        bsv = []
+        for sweep in sweeper:
+            param_name = sweep.parameter.name.lower()
+            base_sweeper_values = [getattr(pulse, param_name) for pulse in sweep.pulses]
+            bsv.append(base_sweeper_values)
+
+        res = self._sweep(qubits, couplers, sequence, options, *sweeper)
+
+        # Reset pulse values back to original values
+        for sweep, base_sweeper_values in zip(sweeper, bsv):
+            param_name = sweep.parameter.name.lower()
+            for pulse, value in zip(sweep.pulses, base_sweeper_values):
+                setattr(pulse, param_name, value)
+
+                if pulse.type is PulseType.READOUT:
+                    res[pulse.serial] = res[pulse.qubit]
```

Comment:
The current representation of the readout pulse serial is
```        
return f"ReadoutPulse({self.start}, {self.duration}, {format(self.amplitude, '.6f').rstrip('0').rstrip('.')}, {format(self.frequency, '_')}, {format(self.relative_phase, '.6f').rstrip('0').rstrip('.')}, {self.shape}, {self.channel}, {self.qubit})"
```
For the mutable temporary sweepers I have at the moment, as the readout pulse parameter is being modified, so too is the serial in every iteration. This makes the final result tied to an accumulation of different readout pulses rather than the original pulse serial.

Its not nice to have the multiple keys, but its a temp workaround for qibocal compatiabilty. This will be removed once I move the sweepers to hardware.

Reply:
Ok, thanks for the explanation.